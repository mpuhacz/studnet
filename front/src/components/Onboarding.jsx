import React, { Component } from 'react';
import Autosuggest from 'react-autosuggest';

const towns = [{
  name: 'Poznan',
}, {
  name: 'Warszawa',
}, {
  name: 'Gdańsk',
}, {
  name: 'Krakow',
}];

const universities = [{
  name: 'Politechnika',
}, {
  name: 'Uniwersytet',
}];

const courses = [{
  name: 'Informatyka',
}, {
  name: 'BioInformatyka',
}, {
  name: 'Matematyka',
}, {
  name: 'Automatyka i Robotyka',
}];

const getTownSuggestions = value => {
  const inputValue = value.trim().toLowerCase();
  const inputLength = inputValue.length;

  return inputLength === 0 ? [] : towns.filter(lang =>
    lang.name.toLowerCase().slice(0, inputLength) === inputValue
  );
};

const getUniversitySuggestions = value => {
  const inputValue = value.trim().toLowerCase();
  const inputLength = inputValue.length;

  return inputLength === 0 ? [] : universities.filter(lang =>
    lang.name.toLowerCase().slice(0, inputLength) === inputValue
  );
};

const getCourseSuggestions = value => {
  const inputValue = value.trim().toLowerCase();
  const inputLength = inputValue.length;

  return inputLength === 0 ? [] : courses.filter(lang =>
    lang.name.toLowerCase().slice(0, inputLength) === inputValue
  );
};

const getSuggestionValue = suggestion => suggestion.name;

const renderSuggestion = suggestion => (
  <span>{suggestion.name}</span>
);

class Onboarding extends Component {
  constructor() {
    super();
    this.state = {
      town: '',
      university: '',
      course: '',
      suggestions: []
    };
  }

  componentWillReceiveProps() {
    if (this.props.location.pathname === '/onboarding' && this.props.preferencesSet) {
      this.props.goToStream();
    }
  }

  onTownChange = (event, { newValue }) => {
    this.setState({
      town: newValue
    });
  };

  onUniversityChange = (event, { newValue }) => {
    this.setState({
      university: newValue
    });
  };

  onCourseChange = (event, { newValue }) => {
    this.setState({
      course: newValue
    });
  };

  onTownSuggestionsFetchRequested = ({ value }) => {
    this.setState({
      suggestions: getTownSuggestions(value)
    });
  };

  onUniversitySuggestionsFetchRequested = ({ value }) => {
    this.setState({
      suggestions: getUniversitySuggestions(value)
    });
  };

  onCourseSuggestionsFetchRequested = ({ value }) => {
    this.setState({
      suggestions: getCourseSuggestions(value)
    });
  };

  onSuggestionsClearRequested = () => {
    this.setState({
      suggestions: []
    });
  };

  onFormSubmit = (e) => {
    e.preventDefault();
    this.props.goToStream(this.state.town, this.state.university, this.state.course);
  }

  render() {
    const { town, course, university, suggestions } = this.state;

    const townInputProps = {
      placeholder: 'Wprowadz miasto',
      value: town,
      onChange: this.onTownChange
    };

    const universityInputProps = {
      placeholder: 'Wprowadz nazwę uczelni',
      value: university,
      onChange: this.onUniversityChange
    };

    const courseInputProps = {
      placeholder: 'Wprowadz kierunek',
      value: course,
      onChange: this.onCourseChange
    };

    return (
      <div className="ui center aligned container xs-margin">
       <div className="ui segment">
          <h1 className="ui header">Witaj w aplikacji StudNET</h1>
          <h5 className="ui header">Aby ułatwic Tobie korzystanie z aplikacji i spersonalizowac tresci, podaj prosze pare informacji o sobie</h5>
          <form className="ui form" onSubmit={this.onFormSubmit}>
            <Autosuggest
              suggestions={suggestions}
              onSuggestionsFetchRequested={this.onTownSuggestionsFetchRequested}
              onSuggestionsClearRequested={this.onSuggestionsClearRequested}
              getSuggestionValue={getSuggestionValue}
              renderSuggestion={renderSuggestion}
              inputProps={townInputProps}
            />
            <Autosuggest
              suggestions={suggestions}
              onSuggestionsFetchRequested={this.onUniversitySuggestionsFetchRequested}
              onSuggestionsClearRequested={this.onSuggestionsClearRequested}
              getSuggestionValue={getSuggestionValue}
              renderSuggestion={renderSuggestion}
              inputProps={universityInputProps}
            />
            <Autosuggest
              suggestions={suggestions}
              onSuggestionsFetchRequested={this.onCourseSuggestionsFetchRequested}
              onSuggestionsClearRequested={this.onSuggestionsClearRequested}
              getSuggestionValue={getSuggestionValue}
              renderSuggestion={renderSuggestion}
              inputProps={courseInputProps}
            />
            <button className="ui button xs-margin" type="submit">
              Gotowe!
            </button>
          </form>
        </div>
      </div>
    );
  }
}

export default Onboarding;
