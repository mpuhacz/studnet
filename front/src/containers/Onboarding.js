import { connect } from 'react-redux';
import Onboarding from '../components/Onboarding';
import { loginAPI } from '../actions/index';
import { push } from 'react-router-redux'

const mapStateToProps = (state, ownProps) => { console.log(state); return {
  preferencesSet: state.reducers.user.preferencesSet,
  isLoading: state.reducers.user.isSettingPreferences,
}};

const mapDispatchToProps = (dispatch, ownProps) => ({
  setUserPreferences: (town, school, course) => dispatch(loginAPI(town, school, course)),
  goToStream: () => dispatch(push('/stream')),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Onboarding);