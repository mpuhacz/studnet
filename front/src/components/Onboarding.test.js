import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Onboarding from './Onboarding';

beforeEach(() => {
  window.location.replace = jest.fn();
});

describe('<Onboarding />', () => {
  it('renders Onboarding without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Onboarding />, div);
  });

  it('renders Onboarding form with correct specification', () => {
    const wrapper = renderer.create(<Onboarding/>);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('fetches course suggestions correctly', () => {
    const wrapper = renderer.create(<Onboarding/>);

    wrapper.getInstance().onCourseSuggestionsFetchRequested({ value: 'Infor' });
    expect(wrapper.getInstance().state).toMatchSnapshot();
  });

  it('fetches university suggestions correctly', () => {
    const wrapper = renderer.create(<Onboarding/>);

    wrapper.getInstance().onUniversitySuggestionsFetchRequested({ value: 'poli' });
    expect(wrapper.getInstance().state).toMatchSnapshot();
  });

  it('fetches town suggestions correctly', () => {
    const wrapper = renderer.create(<Onboarding/>);

    wrapper.getInstance().onTownSuggestionsFetchRequested({ value: 'poz' });
    expect(wrapper.getInstance().state).toMatchSnapshot();
  });
});
