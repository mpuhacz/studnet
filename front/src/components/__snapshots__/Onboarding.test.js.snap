exports[`<Onboarding /> fetches course suggestions correctly 1`] = `
Object {
  "course": "",
  "suggestions": Array [
    Object {
      "name": "Informatyka",
    },
  ],
  "town": "",
  "university": "",
}
`;

exports[`<Onboarding /> fetches town suggestions correctly 1`] = `
Object {
  "course": "",
  "suggestions": Array [
    Object {
      "name": "Poznan",
    },
  ],
  "town": "",
  "university": "",
}
`;

exports[`<Onboarding /> fetches university suggestions correctly 1`] = `
Object {
  "course": "",
  "suggestions": Array [
    Object {
      "name": "Politechnika",
    },
  ],
  "town": "",
  "university": "",
}
`;

exports[`<Onboarding /> renders Onboarding form with correct specification 1`] = `
<div
  className="ui center aligned container xs-margin">
  <div
    className="ui segment">
    <h1
      className="ui header">
      Witaj w aplikacji StudNET
    </h1>
    <h5
      className="ui header">
      Aby ułatwic Tobie korzystanie z aplikacji i spersonalizowac tresci, podaj prosze pare informacji o sobie
    </h5>
    <form
      className="ui form"
      onSubmit={[Function]}>
      <div
        className="react-autosuggest__container">
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-expanded={false}
          aria-haspopup={false}
          aria-owns="react-autowhatever-1"
          autoComplete="off"
          className="react-autosuggest__input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Wprowadz miasto"
          role="combobox"
          type="text"
          value="" />
        <div
          className="react-autosuggest__suggestions-container"
          id="react-autowhatever-1" />
      </div>
      <div
        className="react-autosuggest__container">
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-expanded={false}
          aria-haspopup={false}
          aria-owns="react-autowhatever-1"
          autoComplete="off"
          className="react-autosuggest__input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Wprowadz nazwę uczelni"
          role="combobox"
          type="text"
          value="" />
        <div
          className="react-autosuggest__suggestions-container"
          id="react-autowhatever-1" />
      </div>
      <div
        className="react-autosuggest__container">
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-expanded={false}
          aria-haspopup={false}
          aria-owns="react-autowhatever-1"
          autoComplete="off"
          className="react-autosuggest__input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Wprowadz kierunek"
          role="combobox"
          type="text"
          value="" />
        <div
          className="react-autosuggest__suggestions-container"
          id="react-autowhatever-1" />
      </div>
      <button
        className="ui button xs-margin"
        type="submit">
        Gotowe!
      </button>
    </form>
  </div>
</div>
`;
